src/biokbase/narrative/jobs/appmanager.py

```diff
@@ -466,41 +429,281 @@ def run_app(
+                app once.
+        cell_id: if provided, this should be a unique id for the Narrative cell that's running the app.
+        run_id: if provided, this should be a unique id representing a Narrative cell's knowledge of
+            that job.
+        dry_run: if True, this won't start the job, but return the structure that would be sent to the
+            KBase execution engine.
+
+        Example:
+        --------
+        run_app_batch([{
+            "app_id": "MegaHit/run_megahit",
+            "tag": "release",
+            "version": "1.0.0",
+            "params": [{
+                    "read_library_name" : "My_PE_Library",
+                    "output_contigset_name" : "My_Contig_Assembly"
+            }]
+        }])
+        """
+
+        if not isinstance(app_info, list) or len(app_info) == 0:
+            raise ValueError(
+                "app_info must be a list with at least one set of app information"
+            )
+        batch_run_inputs = list()
+        ws_id = strict_system_variable("workspace_id")
+        batch_params = {"wsid": ws_id}  # for EE2.run_job_batch
+        log_app_info = list()
+        for info in app_info:
+            self._validate_bulk_app_info(info)
+            app_id = info["app_id"]
+            tag = info.get("tag", "release")
+            version = info.get("version")
+            spec = self._get_validated_app_spec(app_id, tag, True, version)
+            for param_set in info["params"]:
+                # will raise a ValueError if anything is wrong or missing
+                # otherwise, will build a set of inputs for EE2.run_job
+                batch_run_inputs.append(
+                    self._build_run_job_params(
+                        spec,
+                        tag,
+                        param_set,
+                        version=version,
+                        cell_id=cell_id,
+                        run_id=run_id,
+                        ws_id=ws_id,
+                    )
+                )
+            log_app_info.append(
+                {
+                    "app_id": app_id,
+                    "tag": tag,
+                    "version": version,
+                    "num_jobs": len(batch_run_inputs),
+                }
+            )
+        log_info = {
+            "app_info": log_app_info,
+            "username": system_variable("user_id"),
+            "wsid": ws_id,
+        }
+        kblogging.log_event(self._log, "run_app_bulk", log_info)
+
+        # if we're doing a dry run, stop here and return the setup
+        if dry_run:
+            return {"batch_run_params": batch_run_inputs, "batch_params": batch_params}
+
+        # We're now almost ready to run the job. Last, we need an agent token.
+        token_name = "KBApp_{}".format(app_id)
+        token_name = token_name[: self.__MAX_TOKEN_NAME_LEN]
+        agent_token = auth.get_agent_token(auth.get_auth_token(), token_name=token_name)
```

Comment:
TODO: factor this into a separate function for run_app, run_app_batch, run_app_bulk,